doubly_linked_list/doubly_linked_list.py

```diff
@@ -25,68 +22,101 @@ def insert_before(self, value):
-
-
 class DoublyLinkedList:
     def __init__(self, node=None):
         self.head = node
         self.tail = node
         self.length = 1 if node is not None else 0
-
     def __len__(self):
         return self.length
-
     """Wraps the given value in a ListNode and inserts it 
     as the new head of the list. Don't forget to handle 
     the old head node's previous pointer accordingly."""
     def add_to_head(self, value):
-        pass
-
+        self.length += 1
+        if not self.head and not self.tail:
+            # Empty list, this is head and tail
+            self.head = self.tail = ListNode(value)
+        else:
+            # We know that the list is populated
+            self.head.insert_before(value)
+            self.head = self.head.prev
     """Removes the List's current head node, making the
     current head's next node the new head of the List.
     Returns the value of the removed Node."""
     def remove_from_head(self):
-        pass
-
+        value = self.head.value
+        self.delete(self.head)
+        return value
     """Wraps the given value in a ListNode and inserts it 
     as the new tail of the list. Don't forget to handle 
     the old tail node's next pointer accordingly."""
     def add_to_tail(self, value):
-        pass
-
+        self.length += 1
+        if not self.head and not self.tail:
+            # Empty list, this is head and tail
+            self.head = self.tail = ListNode(value)
+        else:
+            # We know that the list is populated
+            self.tail.insert_after(value)
+            self.tail = self.tail.nex
     """Removes the List's current tail node, making the 
     current tail's previous node the new tail of the List.
     Returns the value of the removed Node."""
     def remove_from_tail(self):
-        pass
-
+        value = self.tail.value
+        self.delete(self.tail)
+        return value
     """Removes the input node from its current spot in the 
     List and inserts it as the new head node of the List."""
     def move_to_front(self, node):
-        pass
-
+        self.delete(node)
+        self.add_to_head(node.value)
     """Removes the input node from its current spot in the 
     List and inserts it as the new tail node of the List."""
     def move_to_end(self, node):
-        pass
-
+        self.delete(node)
+        self.add_to_tail(node.value)
     """Removes a node from the list and handles cases where
     the node was the head or the tail"""
     def delete(self, node):
-        pass
+        # Planning
+        # If LL is empty
+        if not self.head and not self.tail:
+            print("ERROR: Attempted to delete node not in list")
+            return
+        # If node is head
+        # If node is both
+        elif self.head == self.tail:
+            self.head = None
+            self.tail = None
+        elif node == self.head:
+            self.head = self.head.next
+            node.delete()
+        # If node is tail
```

Comment:
In cases like this your comment is redundant since the python if statement is somewhat self explanatory. Consider focusing your comments on _why_ the code exists rather than what it's doing.